# Notebook 1.1.2: The Falling Penny (Modeling with Computation)

<br>



---



*Modeling and Simulation in Python*

Copyright 2021 Allen Downey, (License: [Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International](https://creativecommons.org/licenses/by-nc-sa/4.0/))

Revised, Mike Augspurger (2021-present)

<br>

---

## Modeling the Velocity of a Penny Falling

Now we're going to create a model to determine the final speed and time required for the penny to fall.

<br>

The statistician George Box famously said "All models are wrong, but
some are useful." The puzzle that modelers need to solve is how to create a model that combines usefulness (accuracy) and wrongness (efficiency) in the right mix: we want a model that gives us a usable answer without wasting time and resources.

<br>

Generally, we start simple and add complexity as necessary.  The model we'll start with is a simple one: in fact, it will look a fair bit like a homework problem from a physics class.  This is an important connection to make, because most of your homework problems in physics or chemistry or biology *are* models--that is, they are simplications of a real world situation designed to help us understand the real world situation better.  You've done more modeling than you even knew!

### Investigation and Abstraction

In your investigation in the exercises of the last notebook (i.e. observing the *Mythbusters* video), you probably decided that drag was an important factor for the velocity of the penny, and you were correct.   But sometimes we implement a model that we know is too simple just to make sure our implementation is successful, and then we iterate on that model to add more complexity.

<br>

We'll do that here.  We're going to reduce our list of possible factors affecting the final velocity of the penny to two factors: gravitational force and height.  We'll assume the effect of all other factors is *negligible* (that is, they have only small effects), and so we will ignore them.  Engineers do this all the time, no matter how much the mathematicians resent it 😏

<br>

So we have completed the first half of abstraction: the selection of our key factors.  The second part of abstraction is defining the *relationships* between the key factors.  In the last notebook, we dropped a penny from a ladder (in our heads, anyway) and hypothesized that the relationship between time and velocity of a falling penny was linear:

$$v=Ct$$

Another form of "investigation" would be to look at the work of previous scientists.  A quick look at an intro physics book would reveal this work, in the form of the *kinematic equations*.  These equations tell us about the relationships between time, height and velocity for an object in free-fall.   

<br>

Happily, these equations tell us that our ladder investigation was valid!  The kinematic equation relating time and velocity in free-fall is:

$$v = g t$$

This model tells us that if the initial velocity is 0, the velocity after time $t$ is a function of the gravitational constant $g$, which at sea level is approximately $g=9.81$ $m/s^2$. 

<br>

A second kinematic equation relates time and distance fallen $x$: 

$$x = g t^2 / 2$$

With a little math magic, we can combine these two equations to find the relationship between our two factors and the final velocity of the penny:

$$ v = \sqrt{2gx}$$

This equation is our abstracted model for the final velocity of a penny-dropped from a height $x$ when drag is negligible.


✅ **Active Reading: Why might it make sense to ignore drag in our model, despite the fact that we know drag is important?**

### Implementation and Validation

"Implementation" in this case is pretty straight-forward: we can just take the values from our particular situation and plug them into our model equation.

<br>

Plugging in a value for $g$ and the height of the Empire State Building ($x = 381$ m), we get $v = 86$ m/s, or about 190 miles per hour.

<br>

That sounds like it could hurt.  We don't need to do much here in terms of validation, either.  Our intuition (aided by the *Mythbusters*) tells us that a penny travelling at 190 mph would quickly be slowed down by drag.  

<br>

This model needs some iteration....

### Iteration

So far, we've just done a pretty simple physics calculation model: no need for a computer simulation yet.

<br>

But as the *Mythbusters* guys showed, this is not a good model: ignoring air resistance leads to major errors. So we want to *iterate* our model: that is, we want to add in some new complication that will make the model more troublesome, but happily more accurate.  It turns out that once
the penny gets to about 29 m/s, the upward force of air resistance
equals the downward force of gravity, so the penny stops accelerating.
This is the *terminal velocity* of the penny in air.

<br>

Once the penny reaches terminal velocity, it doesn't matter how much farther it falls; it hits the sidewalk at about 29 m/s (65 mph).
That's much less than 86 m/s, as the simple model predicts.

<br>

We'll use Python's computational tools in the rest of this chapter to simulate this more complex model to try to find the time it takes for the penny to land.

<br>

---

## Implementation (Second Attempt)

We are now inside the implementation/ validation/ iteration feedback loop!

<br>

We'll start by using Python to compute the results from the previous section.  First we'll create a variable to represent acceleration due to gravity in meters per second squared (m/s$^2$).

In [ ]:
g = 9.8

A *variable* is a name that corresponds to a value. In this example, the name is `g` and the value is the number `9.8`.  When you run the above cell (click the `play` button or press `shift-enter`), nothing seems to happen.  But in fact something does: the code sets the variable `g` equal to `9.8`.

Suppose we let the penny drop for $3.4$ seconds (s).  I'll create a variable to represent this time:

In [ ]:
t = 3.4
t

3.4

Notice that when I add the second line `t`, running the cell produces an output.  This doesn't change the variable at all: it just tells the computer to display the value of `t`.

Now we can compute the velocity of the penny after `t` seconds, and assign the answer to a variable named `v`:

In [ ]:
v = g * t

Python uses the symbol `*` for multiplication.  The other arithmetic operators are `+` for addition, `-` for subtraction, `/` for division, and `**` for exponentiation.

The variable `v` now has been assigned a value.  We can display its value like this:

In [ ]:
v

33.32

After $3.4$ s, the velocity of the penny is about $33$ m/s (ignoring air resistance).  Now let's see how far it would travel during that time.  Remember from above that our equation for distance travelled is $x = g t^2 / 2$:

In [ ]:
x = g * t**2 / 2
x

56.644

It would travel about $56$ m.  Now, going in the other direction, let's compute the time it takes to fall 381 m, the height of the Empire State Building.

In [ ]:
h = 381

For this computation, we need the square root function, `sqrt`, which is provided by a *library* called NumPy, which contains functions related to the manipulation of numbers.

NumPy is built into Colab.  But before we can use it in our notebook,we have to *import* it.  Importing a function means we make it available to use, even though the code for that function is not in this particular notebook:

In [ ]:
from numpy import sqrt

Again using the kinematic equation that we used above, we can write:

In [ ]:
t = sqrt(2 * h / g)
t

8.817885349720552

With no air resistance, it would take about $8.8$ s for the penny to reach the sidewalk.

In [ ]:
v = g * t
v

86.41527642726142

And its velocity on impact would be about $86$ m/s.

### Managing your variables

When you change the contents of a cell, you have to run it again for those changes to have an effect.  If you forget to do that, the results can be confusing, because the code you are looking at is not the code you ran.

<br>

If you ever lose track of which cells have run, and in what order, you should go to the `Runtime` menu and select "Restart & Run All".  Restarting the means that all of your variables get deleted, and running all the cells means all of your code will run again, in the right order.  If you just want to clear the variables without running everything, choose the option to `Restart Runtime`(or press `ctrl-m-period`).

<br>

If you want to see which vriables have been assigned a value, click the `{x}` button on the left side toolbar: this will show you which variables exist currently.  The `type` indicates that kind of variable (`float` is a decimal number, `int` is an integer, etc...).  If you hover over the listed variable, the value assigned to that variable will be displayed.

<br>

Select "Restart & Run All" now and confirm that it runs all of the cells.

### False Precision

Python displays results with about 16 digits, which gives the impression that they are very precise, but don't be fooled.
The numbers we get out are only as good as the numbers we put in.

<br>

For example, the "roof height" of the [Empire State Building is $380$ m](https://en.wikipedia.org/wiki/Empire_State_Building).
We'll use $h=381$ m for this example on the assumption that the observation deck is on the roof and you drop the penny from a 1 meter railing.
But that's probably not right, so we should treat this value as an approximation where only the first two digits are likely to be right.

<br>

If the precision of the inputs is two digits, the precision of the outputs is two digits, *at most*.
That's why, if the output is `86.41527642726142`, I report it as "about 86".

<br>

Just as important, remember that we ignored a lot of factors: humidity, wind speed, air density, etc...  This simplification process increases *uncertainty* and *error*, which are a technical terms that are used to express how confident we are in our results.  Using reasonable significant figures is a first step in trying to express uncertainty.

✅ **Active Reading: Our model tells us that the time required for the drop is 8.817885 s.  Why might "about 10 s" be a better answer?**

### Units and Computation

The computations we just did use numbers without units.
When we set `h=381`, we left out the meters, and when we set `g=9.8`, we left out the meters per second squared.
And, when we got the result `v=86`, we added back the meters per second.  Leaving units of out computation is common practice, but it tends to cause errors, including [the very expensive failure of the Mars Climate Orbiter in 1999](https://en.wikipedia.org/wiki/Mars_Climate_Orbiter).

<br>

We have two choices.  We could us a units library such as [Pint](https://pint.readthedocs.io/).
This helps to find errors in the code because it recognizes unit mistakes, but I (and most programmers) find it unwieldy: it requires too much attention and mucks up the code.

<br>

Our second choice is to always use base units.  In the metric system, if we always put our values in $kg$ (kilogram),$m$ (meter), and $s$ (seconds), as well as other more complex base units like Newtons and Watts, we know that my answer will be in base units (presuming I set up my equations correctly).  This often requires using exponents, since a given simulation might include millions of seconds or a small fraction of a kg, but it allows us to generally avoid unit conversions. This is the approach we'll use in this textbook.

<br>

Unfortunately, life does not always provide you with metric base units.  So in some cases, you need to convert data from or solutions or into more common units (like "miles per hour").  In those cases, you want to be extra careful in your conversions: always test your conversions by printing out a test case and making sure the code is converting correctly.

### Error Messages

In mathematical notation, we can write an equation like $v = g t$ and it's understood that we are multiplying $a$ and $t$.
But that doesn't work in Python.  If you put two variables side-by-side, like this:

```
v = g t
```

you'll get a *syntax error*, which means that something is wrong with the structure of the program.
Try it out in the next cell so you see what the error message looks like.

In [ ]:
v = gt

The error messages you get from Python can be big and scary (not this one), but if you read them carefully, they contain a lot of useful information.

<br>

The last line usually tells you what type of error happened, and sometimes additional information, so you might want to start from the bottom and read up.  Also look for the arrows (---->) in the error message.  Sometimes there are multiple arrows, but one of them will point to the line in the code that led to the error.

<br>

The previous lines are a *traceback* of what was happening when the error occurred.  The first section of the traceback shows the code you wrote.  The following sections are often from Python libraries.

<br> 



---



## Summary and Exercises

As a first example, we modeled a penny dropped from the Empire State building, including gravity but ignoring air resistance.  In the exercises, you'll iterate the model a couple times in an attempt to improve its accuracy.

<br>

---

### Exercise 1

In this chapter we used the `sqrt` function from the NumPy library.  NumPy also provides a variable named `pi` that contains an approximation of the mathematical constant $\pi$.
We can import it like this:

In [ ]:
from numpy import pi
pi

NumPy provides other functions we'll use, including `log`, `exp`,`square`,`power`, `sin`, and `cos`.  To find out how to use these functions, see https://numpy.org/doc/stable/reference/routines.math.html

Import `sin` and `cos` from NumPy and compute

$$sin^2 (\pi/4) + cos^2 (\pi/4)$$

Note: A mathematical identity tells us that the answer should be $1$.  There are are multiple ways to do this, including using `power` or `square` from the NumPy library (remember to import them).  Try to find a couple ways to get the correct answer.  

In [ ]:
# Complete your solution here.



Did you notice that you often do not get exactly $1$?  This is the difference between *analytical math* (which tells us that the answer is exactly $1$) and *computational math* (which routinely has (generally) small rounding errors).


### Exercise 2

Suppose instead of dropping the penny, you throw it downward at its terminal velocity, $29$ m/s.  How long would it take to fall $381$ m?   (Hint: If we know a constant velocity and distance travelled, we can find the time.  Consider a similar question: if I have to drive 180 miles to Chicago, and drive 60 mph, how long will it take me?  How do I get that answer?)

Find the answer using Python tools below.

In [ ]:
# Complete the solution here.  The penny would fall at a constant velocity of 29 m/s.  


### Exercise 3

We started with the question, "What is the velocity of the penny when it hits the ground?"  Our discovery that the terminal velocity is 29 m/s (65 mph) seems to answer that question: the velocity will be 29 m/s!  

<br>

Why is that a problematic answer?  Why do the *Mythbusters* guys say the terminal velocity is "between 35 and 65 mph" (and [this site](https://mechse.illinois.edu/news/blogs/lethal-pennies-and-other-misconceptions) gives a slightly different answer)?  What might affect the actual velocity?

### Exercise 4

So far we have considered two models of a falling penny:

* If we ignore air resistance, the penny falls with constant acceleration, and we can compute the time to reach the sidewalk and the velocity of the penny when it gets there.

* If we take air resistance into account, and drop the penny at its terminal velocity, it falls with constant velocity. 

Now let's consider a third model that includes elements of the first two: let's assume that the acceleration of the penny is `a` until the penny reaches $29$ m/s, and then $0$ m/s afterwards.  What is the total time for the penny to fall $381$ m?

This is a two phase simulation, because we have to break the problem into 2 time periods: when it is accelerating downward, and when its acceleration is zero.  To solve, answer the following 3 questions:

1. How long would the penny take to reach $29$ m/s with constant acceleration `g`.  When an object is accelerating at $9.8  m/s^2$, that means every second its velocity increases by $9.8  m/s$.
2. How far would it fall during that time?
3. How long would it take to fall the remaining distance with constant velocity $29$ m/s?

Suggestion: Assign each intermediate result to a variable with a meaningful name.   Here are the kinematic equations from the start of this notebook:

$$v = g t$$

$$x = g t^2 / 2$$

$$ v = \sqrt{2gx}$$

In [ ]:
# Establish the unchanging variables (these are called "parameters").  Use base metric units.


In [ ]:
# Find how long it will take the penny to reach terminal velocity.  Assign that value to a new variable.


In [ ]:
# Find how far it will fall while it is reaching terminal velocity.  Use kinematic equations.


In [ ]:
# Find how long it will take to fall the remaining distance at constant velocity


In [ ]:
# Determine the total time.  Assign this to a variable, and output the value of this variable.


According to this model, it takes almost 15 seconds to fall to the ground, which is more than our first model (8.8 s) as well as our second model (13.1 s).